### PR TITLE
chore: cleanup types

### PR DESCRIPTION
**Pull request by nickofthyme**
## Details

- remove unnecessary `// @ts-ignore` comments
- add descriptions to legit `// @ts-ignore` comments
- add [rule](https://github.com/typescript-eslint/typescript-eslint/blob/v4.31.1/packages/eslint-plugin/docs/rules/ban-ts-comment.md) to error on all ts-comments with no description
- remove `global.d.ts` in favor of `declarations/`
- fix type issue related to `jest-extended` and custom matchers
- fix types around `sortSeriesBy` hidden setting
- fix type errors around required `domain` options
- fix type errors around union brush events
- remove deprecated [`msSaveBlob`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/msSaveBlob) usage
- add step to ci to check `yarn typecheck:all` as type errors somehow snuck into `master`

### Eslint changes

Before eslint would **WARN** ⚠️ you when using `// @ts-ignore`, `// @ts-expect-error`, `// @ts-nocheck` or `// @ts-check`. This caused some to ignore the eslint warning but now it is explicit why we are ignoring a given line.

Now it will **ERROR** ❌  when using any of the above _without_ a description, otherwise it will not complain.

Image 2021-09-23 at 12 48 21 PM

### Checklist

- [x] New public API exports have been added to `packages/charts/src/index.ts`

### CODE REVIEW

File: storybook/stories/area/13_band_area.story.tsx [hunk 1]
```diff
@@ -50,7 +50,11 @@ export const Example = () => {
       />
       <Axis
         id="left"
-        domain={{ fit }}
+        domain={{
+          min: NaN,
+          max: NaN,
+          fit,
+        }}
```
Comment (monfera) [inline storybook/stories/area/13_band_area.story.tsx:57]: Nice catch! For some reason nothing brought it up

File: storybook/stories/bar/44_test_single_histogram.story.tsx [hunk 2]
```diff
@@ -27,6 +27,8 @@ export const Example = () => {
   const formatter = timeFormatter(niceTimeFormatByDay(1));
 
   const xDomain = {
+    min: NaN,
+    max: NaN,
```
Comment (monfera) [inline storybook/stories/bar/44_test_single_histogram.story.tsx:31]: Wow so many of them 😄

File: storybook/stories/interactions/17_png_export.story.tsx [hunk 3]
```diff
@@ -51,17 +51,12 @@ export const Example = () => {
...
-        document.body.appendChild(link);
-        link.click();
-        document.body.removeChild(link);
-    }
+    const link = document.createElement('a');
+    link.download = fileName;
+    link.href = snapshot.blobOrDataUrl;
+    document.body.appendChild(link);
+    link.click();
+    document.body.removeChild(link);
```
Comment (monfera) [inline storybook/stories/interactions/17_png_export.story.tsx:59]: IE removal ftw!

File: storybook/stories/line/2_w_axis.story.tsx [hunk 4]
```diff
@@ -26,7 +26,13 @@ const dateFormatter = timeFormatter(niceTimeFormatByDay(1));
...
 export const Example = () => (
   <Chart>
-    <Settings xDomain={{ max: 2 }} baseTheme={useBaseTheme()} />
+    <Settings
+      xDomain={{
+        min: NaN,
+        max: 2,
+      }}
+      baseTheme={useBaseTheme()}
+    />
```
Comment (monfera) [inline storybook/stories/line/2_w_axis.story.tsx:35]: It doesn't matter at all in story files and I should be the last person to bring it up with the NaNs, just wanted to mention that in regular source code it's nice to retain a single line for an object or sub-object if props aren't super numerous or deeply nested (in which case the linter breaks the lines anyway)

File: storybook/stories/area/13_band_area.story.tsx (hunk 1 above)
Comment (nickofthyme) [reply storybook/stories/area/13_band_area.story.tsx:57]: Yeah I'm not sure why, which is why I added `yarn typecheck:all` to the ci checks.

File: storybook/stories/line/2_w_axis.story.tsx (hunk 4 above)
Comment (nickofthyme) [reply storybook/stories/line/2_w_axis.story.tsx:35]: Yeah I was copy pasting from a multiline so some got push to the next line. We can fix this later.

**Comment by nickofthyme** [comment]
Jenkins test this please

**Review by rshen91 — APPROVED** [review]
I like these changes! LGTM! Thanks for going through and cleaning up unneeded @ts-ignore comments; I especially like the change to have comments when using @ts-ignore

File: declarations/@types/jest.d.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * Final Matcher type with `this` and `received` args removed from jest matcher function
+ */
+type MatcherParameters<T extends (this: any, received: any, ...args: any[]) => any> = T extends (
+  this: any,
+  received: any,
+  ...args: infer P
+) => any
+  ? P
+  : never;
```
Comment (monfera) [inline declarations/@types/jest.d.ts:20]: Great, we should use more and more `never`, eg. tying up the loose ends of `switch/case` statements or anything that branches over the full union of something (eg. the ending of `rawChildNodes`)

File: storybook/stories/area/13_band_area.story.tsx (hunk 1 above)
Comment (monfera) [reply storybook/stories/area/13_band_area.story.tsx:57]: Great! What's the local equivalent, just running `yarn typecheck:all`? Does it remove the need for running the API extraction, esp. if the API doesn't change?

File: storybook/stories/line/2_w_axis.story.tsx (hunk 4 above)
Comment (monfera) [reply storybook/stories/line/2_w_axis.story.tsx:35]: These are perfectly OK here

**Review by monfera — APPROVED** [review]
Great improvement 🎉  My current review is cursory, will report any oddities, if any, down the road

**Pull request merged**